Duotone: Only output Duotone filters that are used

scruffian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This is an attempt at fixing https://github.com/WordPress/gutenberg/issues/38299. The idea is that by outputting only the SVGs that are used, rather than all of the presets, we can remove a lot of unused SVGs on the frontend.

## Why?
Less code will make the page load faster.

## How?
1. Create a new WP_Gutenberg class.
2. Add an action to the WP_Theme_JSON class to make is possible to hook onto declarations.
3. Save a list of the CSS variables that duotone uses to reference SVGs, in a static array on the new class.
4. Add a filter to the WP_Theme_JSON class to make is possible to output SVGs from another class.
5. Output the SVGs that are saved in the static array created in step 1.

## Testing Instructions
1. Switch to a theme that has a default duotone applied to all blocks (you can use the Sherbet variation of TT3)
2. Add an image to a post
6. Confirm that the image has a duotone applied
7. Check the page source and check that none of the default duotone SVGs are being output.

### Testing Instructions for Keyboard
For step 3 above check that the source contains an SVG with an ID which is the same as the id that the CSS on an image references in its `filter` property.
For step 4 check that there are no other SVGs with filters in this format on the page:
`<filter id="wp-duotone-*">`

scruffian committed:
Duotone: Only output Duotone filters that are used

scruffian commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2266,6 +2266,42 @@ private static function get_block_nodes( $theme_json, $selectors = array() ) {
 		return $nodes;
 	}
 
+	private function gutenberg_get_preset_from_declaration( $declaration ) {
```
This is a big function that has a lot of code duplicated from `get_svg_filters`. If we can't find a better way to get the preset then we should at least extract the common code from these two functions into a helper.

scruffian commented on lib/compat/wordpress-6.2/script-loader.php:
```diff
@@ -191,3 +191,7 @@ function gutenberg_enqueue_global_styles_custom_css() {
+
+
+remove_action( 'wp_body_open', 'wp_global_styles_render_svg_filters' );
+remove_action( 'in_admin_header', 'wp_global_styles_render_svg_filters' );
```
This removes the call that adds the SVGs for the duotone presets to the body of the page.

scruffian commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2425,8 +2461,34 @@ function( $pseudo_selector ) use ( $selector ) {
+				$filter_preset = $this->gutenberg_get_preset_from_declaration( $declarations[ $index ] );
+				$filter_svg = gutenberg_get_duotone_filter_svg( $filter_preset );
+
+				add_action(
```
One we have the SVG we then add it to the body of the page. This code also lives in block-supports/duotone.php so we should refactor to share the code.

scruffian commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2425,8 +2461,34 @@ function( $pseudo_selector ) use ( $selector ) {
 		foreach ( $declarations as $index => $declaration ) {
 			if ( 'filter' === $declaration['name'] ) {
-				unset( $declarations[ $index ] );
+				$filter_preset = $this->gutenberg_get_preset_from_declaration( $declarations[ $index ] );
```
This is the one area that needs a lot more attention. What we have is a filter preset of the form `var(--wp--preset--duotone--(.*))`. What we need is a preset object from theme.json to get the settings so we can output the correct SVG. I think there must be a better way of getting the preset than using the slug like this.

ajlende commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -2425,8 +2461,34 @@ function( $pseudo_selector ) use ( $selector ) {
+							 * Simply accessing el.offsetHeight flushes layout and style
+							 * changes in WebKit without having to wait for setTimeout.
+							 */
+							printf(
+								'<script>( function() { var el = document.querySelector( %s ); var display = el.style.display; el.style.display = "none"; el.offsetHeight; el.style.display = display; } )();</script>',
+								wp_json_encode( $selector )
+							);
```
It would be nice if we could avoid the Safari hack in as many cases as possible. It exists to fix this problem. (#37954)

![broken safari duotone](https://lendedotblog.files.wordpress.com/2023/01/duotone-safari.png)

ajlende commented:
I feel a bit weird about adding a side-effect to `get_styles_for_block`. It's the sort of function that could end up getting called elsewhere outside of rendering which could cause issues.

I wonder if we could do something similar to this function that just returns the string of SVGs and then call it in `gutenberg_enqueue_global_styles` or keep the `wp_body_open` action?

scruffian committed:
simplify

scruffian commented on lib/compat/wordpress-6.2/script-loader.php:
```diff
@@ -191,3 +191,38 @@ function gutenberg_enqueue_global_styles_custom_css() {
+ *
+ * @since 5.9.1
+ */
+function gutenberg_global_styles_render_svg_filters() {
```
This is just a copy of `wp_global_styles_render_svg_filters`

scruffian commented on lib/compat/wordpress-6.2/get-global-styles-and-settings.php:
```diff
@@ -255,3 +255,45 @@ function _gutenberg_add_non_persistent_theme_json_cache_group() {
+ *
+ * @return string
+ */
+function gutenberg_get_global_styles_svg_filters() {
```
This is just a copy of `gutenberg_get_global_styles_svg_filters` with the reference to the `WP_Theme_JSON_Resolver` class updated to `WP_Theme_JSON_Resolver_Gutenberg`.

scruffian committed:
just compare the CSS variables

scruffian committed:
remove unused code

scruffian committed:
remove unconnected change

scruffian committed:
fix linter

scruffian committed:
Fix linter

getdave commented:
As I see it the problem people have is that we're dumping a tonne of SVG markup on the page when those SVGs may not be used. So what we need is a way to retain the ability to _select_ from all of the presets but only _load_ the SVGs for the presets that are in use.

This PR partly addresses this but might leave us in [a situation where the available presets are not presented/available to the user](https://github.com/WordPress/gutenberg/pull/48318) because they were not currently in use on the given page.

Perhaps a different approach could be to 

- allow all presets to be defined and availanle as CSS custom props
- only inline SVGs that are currently in use
- lazy load any SVGs that are needed as a result of users selecting a new preset

scruffian commented:
> This PR partly addresses this but might leave us in https://github.com/WordPress/gutenberg/pull/48318 because they were not currently in use on the given page.

I don't think this is true. The presets will continue to be presented to the user, they just might not be output in the frontend.

scruffian committed:
use actions and filters to make the approach more generic

scruffian commented:
I updated the approach here to use actions and filters. I know that this comes with a maintenance burden, but the benefit in this case is that it will allow other plugins to provide their own filters. cc @oandregal

scruffian committed:
remove unused code

jeryj commented:
> I updated the approach here to use actions and filters. I know that this comes with a maintenance burden,

What is the extra maintenance burden with using actions and filters here?

scruffian commented:
> What is the extra maintenance burden with using actions and filters here?

Once an action / filter is in core we have to support it indefinitely.

jeryj commented on lib/block-supports/duotone.php:
```diff
@@ -519,3 +519,44 @@ static function () use ( $filter_svg, $selector ) {
+		$duotone_preset_css_var = WP_Theme_JSON_Gutenberg::get_preset_css_var( array( 'color', 'duotone' ), $duotone_preset['slug'] );
+
+		// Only output the preset if it's used by a block.
+		if ( in_array( $duotone_preset_css_var, WP_Duotone::$duotone_presets, true ) ) {
```
I'm on Theme Twenty Twenty Two and getting a fatal error at this line due to `WP_Duotone::$duotone_presets` being `null`. I haven't figured out why it's getting set to `null` yet, but adding an `is_array( WP_Duotone::$duotone_presets ) &&` check does fix the error.

scruffian replied:
Thanks. It's null because TT2 doesn't set a duotone by default.

jeryj committed:
Default static $duotone_preset to empty array

The in_array check on WP_Duotone::$duotone_presets causes a fatal error if there are no filter declarations present in the theme json.

scruffian committed:
Merge branch 'update/output-only-used-filters' of github.com:WordPress/gutenberg into update/output-only-used-filters

scruffian commented:
One problem with the approach here is that it still outputs the CSS variables for the filters that aren't being used.

scruffian committed:
move the place where we output the css for the preset

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -519,3 +519,47 @@ static function () use ( $filter_svg, $selector ) {
+		// Only output the preset if it's used by a block.
+		if ( in_array( $duotone_preset_css_var_with_wrapper, WP_Duotone::$duotone_presets, true ) ) {
+			// Output the CSS for the preset.
+			// I think we should do this differently, but I'm not sure how.
+			$filters .= '<style>body{' . $duotone_preset_css_var . ':'. gutenberg_get_duotone_filter_property( $duotone_preset ) . ';}</style>';
```
It would be nice if it could still be handled by `value_func` in `WP_Theme_JSON_Gutenberg` instead of moving it here.

Maybe we have to think a little more about where the filters/actions belong in that code? Could it be more useful in `gutenberg_add_global_styles_for_blocks`—modifying the `$metadata` passed into `get_styles_for_block`?

draganescu replied:
I think @ajlende has a point. Somehow appending that preset CSS code doesn't feel it belongs in `get_filter_svg` and since it's a preset and WP Theme json gives us a way to add callback for presets then we should use that.

getdave commented:
> One problem with the approach here is that it still outputs the CSS variables for the filters that aren't being used.

I don't see that as a major problem and not one you need to tackle here. Getting rid of the SVG markup is the main thing.

oandregal commented:
> I updated the approach here to use actions and filters. I know that this comes with a maintenance burden, but the benefit in this case is that it will allow other plugins to provide their own filters. cc @oandregal

Thanks for the ping. Not super fan of adding a filter to a private method/class, as it prevents us from modifying that private unit going forward (it makes it effectively public and part of a flow we should not change). What other options do we have?

Is allowing plugins to modify the theme.json schema something we need? I'd decouple that decision from what this PR is trying to achieve, as it comes with a different set of trade-offs.

scruffian commented:
Closing in favour of https://github.com/WordPress/gutenberg/pull/48374
